doc/proposals/metering-operator-metrics.md

```diff
@@ -0,0 +1,99 @@
+
+```
+crd_kind_info{namespace="namespace",crdkind="instance-name"} 1
+```
+
+example metric for the memchached-operator would look like this:
+
+```
+memcached_info{namespace="default",memcached="example-memcached"} 1
+```
+
+### Existing solutions
+
+Either solutions makes use of Kubernetes list/watch to populate a Prometheus metrics registry, only that kube-state-metrics solution implements its own registry for performance reasons and the client_golang approach makes use of the canonical metrics registries.
+
+#### kube-state-metrics based
+
+[kube-state-metrics](https://github.com/kubernetes/kube-state-metrics#overview) can be used because it solves exactly the same problem but it does it for upstream known resources. The operator-sdk can re-use its functionality do perform the same thing with custom resources.
+
+```go
+// NewCollectors returns a collection of metrics in the namespaces provided, per the api/kind resource.
+// The metrics are registered in the custom generateStore function that needs to be defined.
+func NewCollectors(
+client *Client,
+	namespaces []string,
+	api string,
+	kind string,
+	metricsGenerator func(obj interface{}) []*metrics.Metric) (collectors []*kcoll.Collector)
+```
+
+```go
+// ServeMetrics takes in the collectors that were created and port number on which the metrics will be served.
+func ServeMetrics(collectors []*kcoll.Collector, portNumber int) {
+
+```
+
+Note: Due to taking advantage of kube-state-metrics functions and interfaces we cannot use the prometheus/client_golang and we need to register it in the same way as kube-state-metrics does, and expose the `/metrics` and serve it on a port (port `:8389/metrics` for example). For that we will need to also create a Service object or rather update the current Service object.
+
+#### Using Prometheus client_golang
+
+The advantage of using the Prometheus native [client-golang](https://github.com/prometheus/client_golang) is we can use the already existing metrics registry and therefore not need to serve another http server for the metrics as we would need to in the first solution. 
+
+Another thing to note is that client_golang is also a known interface and the user can make use of the rest of the Prometheus ecosystem. 
+
+The reason why kube-state-metrics implemented its own client is for performance reasons. It is unclear at this time if the operator-sdk will need to handle similar load. Should this be the case and the users need the performance we would eventually need to switch in the same way as kube-state-metrics had to.
+
+The architecture in the operator-sdk would look similar, but having its own collectors and store functions similar to how they exist in kube-state-metrics, but taking advantage of client_golang components.
```

Comment:
Just for clarification, `client_golang` is Prometheus client_golang or Kubernetes client-golang?

Reply:
Yes, client_golang refers to the Prometheus go client. The Kubernetes one is called client-go, but maybe I should refer to Prometheus client_golang as Prometheus client or Prometheus library instead.